src/ice_match/matchers/fly_matcher.py

```
@@ -47,48 +54,117 @@ def find_matches(self, pool_manager: UnmatchedPoolManager) -> List[MatchResult]:
+    def _find_fly_patterns_by_month_grouping(
+        self, trades: List[Trade]
+    ) -> List[List[Trade]]:
+        """Find fly patterns using 3-SUM variant approach - O(m³ + n²) complexity.
+
+        This optimized implementation:
+        1. Pre-builds quantity indexes for O(1) middle leg lookups
+        2. Iterates through month triplets (O(m³), m is typically small)
+        3. Uses hash maps for efficient quantity-based trade lookups
+        4. Applies expensive validations last (fail-fast principle)
+
+        Much more efficient than the previous 6-nested-loop O(n × m³) approach.
```

Ruff (0.12.2)

Docstring contains ambiguous `×` (MULTIPLICATION SIGN). Did you mean `x` (LATIN SMALL LETTER X)?

(RUF002)